src/python/pants/engine/unions.py

```diff
@@ -48,11 +49,17 @@ def non_member_error_message(subject):
     )
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass(unsafe_hash=True)
 class UnionMembership:
-    union_rules: Dict[Type, OrderedSet[Type]]
+    union_rules: FrozenDict[Type, FrozenOrderedSet[Type]]
+
+    def __init__(self, union_rules: Mapping[Type, Iterable[Type]]) -> None:
```

Comment:
oh this is a good pattern

Reply:
I love the pattern. And I love how MyPy makes it intuitive to understand.

Generally, I try to have parameter types be the most flexible possible, but return types the most concrete possible.